### DISCUSSION TITLE
Welcome to Approach Discussions!

### DISCUSSION THREAD

**Discussion by TheApproach** [TheApproach/Approach#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by AbiodunCreatives** [comment]
I'm Abiodun, a Frontend Developer, currently sharpening my skill set to be able to make meaning contributions in the tech space, I'm a very curious type of person, willing to know about alot of things especially technology wise, would be cool to know more about the community and how I can be of help to grow the community.

**Reply by TheApproach** [reply]
If you are interested in working on growing the community, that would be very helpful. Approach has a different way of doing many things in order to have many separations-of-concerns all form one big state machine. This is being used at https://getsuiteux.com to run our cloud, page builder, user accounts, internal intranet, clusters.. all the things a small enterprise needs.

We want to make an open source cloud with the same capabilities, that anyone can fork and be able to share micro-libraries in something like Wordpress's plugin marketplace eventually. 

While the backend and most of the server automations are owned by this project; much of the front end UI had to be left behind as property of our gracious sponsors. Same for the Resource layer which had specialized database connectors.

---

Long story short; whatever skills you already know: there is surely a place to jump in. If you know HTML and PHP, learning how the Render\ namespace uses the Node class as the foundation for XML and HTML classes is a great start. 

If you are more comfortable making UI as static vanilla HTML/CSS we have the Imprint system which will convert your HTML to compiled trees for you. We can attach web intents into HTML attributes and slice things into Components which get their own APIs just by creating HTML templates. Similarly, if you are working with another format than HTML/XML we can explore how those pathways work.

Writing tests, documentation, feedback and UI design are helpful to get the ball rolling.
I or one of the people on our team can definitely help anyone start translating skills to work in Approach.

Once we get the frontends up in the open, you will be able to spin up your own enterprise and roll your own low code platform, piece of cake. Unfortunately things have stalled as my design and technical writing skills are quite limited compared to library engineering.

**Comment by TheApproach** [comment]
![orchestra_sample](https://github.com/TheApproach/Approach/assets/115773679/89c3b630-88ff-4dc1-ab7c-e64c7cd1929a)
Here is a very high level overview of one possible Approach Orchestra

Orchestras are made of multiple "Ensembles" which are essentially clusters. Each ensemble has 3 different ways of addressing it
- by infrastructure, generally the type of machine, instrument, server being used and an id
- by project (the normal way you name internal servers)
- by layer, all code in an Approach project are formed out of building blocks. each layer is a stream timeline made of the lower layers. permissions, roles, etc live in here

Each Ensemble is made of some number of servers, or what we tend to call Instruments.
Within each Instrument exist your project, organized into namespaces that match different parts of the pipeline

When setup properly, someone with very little knowledge of Approach can add or change small, familiar things like HTML patterns and get big results, such as SaaS APIs. Throw in our data and settings mapper backend, it becomes possible to do many things from an IT operations perspective that usually would need hands on coding.

**Comment by TheApproach** [comment]
As you can see there are many things which are currently in the form of a 'brain trust' -- things we simply know about the system but have no proper documentation, user guides, examples or community feedback. Some Codeception tests already exist but we certainly need more on that end -- and most of all to recreate the frontends.

![paradigm layers](https://github.com/TheApproach/Approach/assets/115773679/ba33bd11-d2c7-4de4-9044-19792a3070f4)

Here is a chart of the Approach layers. 

Each layer is a building block of the layer above it. Each layer has a particular class interface and concerns.
The canonical example of the basic "Render" layer are the XML and HTML formats.

For example, this HTML pattern file
![pattern_file_sample](https://github.com/TheApproach/Approach/assets/115773679/da3ff86f-bbde-489d-9ad6-b19be2405c40)

is parsed into this autogenerated code:
![minted_pattern_sample](https://github.com/TheApproach/Approach/assets/115773679/e34b94c4-8c0c-46d3-a96d-dc508cb84b7b)


But you can equally do things such as  [found here](https://github.com/TheApproach/Approach/blob/master/tests/Unit/RenderCest.php) by hand.

This gives developers who know HTML an entry point to PHP that follows the same structure as the output media they are making, but allows them to inject control structures easily such as if statements, loops, wrappers, etc. 

Components are similar to MVC controllers in some ways and Views in some ways. They take in a resource map and bind values into a Minted Pattern's tokens, like seen above. Compositions are made of Components, Resources and Render nodes and are basically your URL edge.

Long story short; we can usually match by their current skillset + what skills do they want to work toward.
